astropy/modeling/tests/test_quantities_parameters.py

```diff
@@ -0,0 +1,476 @@
+    assert exc.value.args[0] == ('TestModel.__init__() requires parameter \'b\' to '
+                                 'be in units equivalent to Unit("m") (got Unit("s"))')
+
+
+def test_parameter_set_equivalency():
+    """
+    Test that we can set equivalencies on an existing model.
+    """
+    g = Gaussian1D(1 * u.Jy, 3 * u.m, 0.1 * u.nm)
+
+    with pytest.raises(UnitsError) as exc:
+        g.mean = 3 * u.Hz
+    assert exc.value.args[0] == ("The 'mean' parameter should be given as a "
+                                 "Quantity with units equivalent to m")
+
+    g.mean.equivalencies = u.spectral()
+
+    g.mean = 3 * u.Hz
+
+    # Note that we can't use this for equivalencies that rely on the values at
+    # which the model is evaluated (such as brightness_temperature which
+    # depends on spectral coordinate) because we won't know what values to use
+    # until evaluation. See the test_output_units_equivalencies_with_parameters
+    # test for how we can deal with this during evaluation.
+
+
+def test_parameter_change_unit():
+    """
+    Test that changing the unit on a parameter works as expected (units can
+    only be changed to a compatible unit).
+    """
+
+    g = Gaussian1D(1, 1 * u.m, 0.1 * u.m)
+
+    # Setting a unit on a unitless parameter should not work
+    with pytest.raises(ValueError) as exc:
+        g.amplitude.unit = u.Jy
+    assert exc.value.args[0] == ("Cannot attach units to parameters that were "
+                                 "not initially specified with units")
+
+    # Changing to an equivalent unit should work
+    g.mean.unit = u.cm
+
+    # But changing to another unit should not
+    with pytest.raises(UnitsError) as exc:
+        g.mean.unit = u.Hz
+    assert exc.value.args[0] == ("Cannot set parameter units to Hz since it is "
+                                 "not equivalent with the original units of cm")
+
+    # Now we test a more complex example with equivalencies
+    class TestModel(Model):
+        a = Parameter(default=1.0, unit=u.m, equivalencies=u.spectral())
+        @staticmethod
+        def evaluate(x, a):
+            return x
+
+    model = TestModel()
+
+    # This should work as before
+    model.a.unit = u.cm
+
+    # This should now work because of the equivalency
+    model.a.unit = u.Hz
+
+    # But this should still not work
+    with pytest.raises(UnitsError) as exc:
+        model.a.unit = u.Jy
+    assert exc.value.args[0] == ("Cannot set parameter units to Jy since it is "
+                                 "not equivalent with the original units of Hz")
+
+
+def test_parameter_set_value():
+    """
+    Test that changing the value on a parameter works as expected.
+    """
+
+    g = Gaussian1D(1 * u.Jy, 1 * u.m, 0.1 * u.m)
+
+    # To set a parameter to a quantity, we simply do
+    g.amplitude = 2 * u.Jy
+
+    # If we try setting the value, we need to pass a non-quantity value
+    # TODO: determine whether this is the desired behavior?
```

Comment:
does it make sense to be able to set the value here, surely either you should a) not have a unit and be setting whole parameter as a float or b) have a unit and have to use Quantities to be explicit about things?